File: comm/gauss_elim_gf2.cpp
/*
 * Find an assignment (result) s.t. xor_i a[i] * result[i] = b (in GF(2))
 * Returns true if such an assignment was found.
 * Verified by: yukicoder No.460 (http://yukicoder.me/submissions/150485)
 */
bool gauss_elim_gf2_i64(vector<ll> a, ll b, vector<bool> &result) {
  int n = a.size();
  int c = 0;
  vector<int> revmap;
  REP(r, 0, 64) {
    if (c >= n) {
      break;
    }
    int c2 = -1;
    REP(i, c, n) {
      if (a[i] & (1LL << r)) {
	c2 = i;
	break;
      }
    }
    if (c2 < 0) {
      revmap.push_back(-1);
      continue;
    }
    if (c != c2) {
      swap(a[c], a[c2]);
    }
    ll rm = a[c] & -(1LL << r) << 1;
    a[c] ^= rm;
    REP(k, c + 1, n) {
      if (a[k] & (1LL << r)) {
	a[k] ^= rm;
      }
    }
    if (b & (1LL << r)) {
      b ^= rm;
    }
    revmap.push_back(c);
    c++;
  }
  // recover
  int rank = revmap.size();
  result.assign(n, false);
  for (int i = rank - 1; i >= 0; --i) {
    if (b & 1LL << i) {
      int c = revmap[i];
      if (c < 0) {
	return false;
      }
      b ^= a[c];
      result[c] = true;
    }
  }
  return b == 0;
}

File: comm/gauss_elim_gf2.rs
/*
 * Find an assignment (result) s.t. xor_i a[i] * result[i] = b (in GF(2))
 * Returns Some if such an assignment was found, None otherwise.
 */
fn gauss_elim_gf2_i64(basis: &[i64], mut b: i64) -> Option<Vec<bool>> {
    let n = basis.len();
    let mut a = basis.to_vec();
    let mut c = 0;
    let mut revmap = Vec::new();
    let w = 64; // i64's size
    for r in 0 .. w {
        if c >= n {
            break;
        }
        let mut c2 = None;
        for i in c .. n {
            if (a[i] & 1 << r) != 0 {
                c2 = Some(i);
                break;
            }
        }
        if c2 == None {
            revmap.push(None);
            continue;
        }
        a.swap(c, c2.unwrap());
        let rm = a[c] & -(1 << r) << 1;
        a[c] ^= rm;
        for k in c + 1 .. n {
            if (a[k] & 1 << r) != 0 {
                a[k] ^= rm;
            }
        }
        if (b & 1 << r) != 0 {
            b ^= rm;
        }
        revmap.push(Some(c));
        c += 1;
    }
    // recover
    let rank = revmap.len();
    let mut result = vec![false; n];
    for i in (0 .. rank).rev() {
        if (b & 1 << i) != 0 {
            match revmap[i] {
                None => return None,
                Some(c) => {
                    b ^= a[c];
                    result[c] = true;
                },
            }
        }
    }
    if b == 0 {
        Some(result)
    } else {
        None
    }
}
